app/src/main/java/pm/gnosis/heimdall/ui/safe/connect/Connect2FaStartActivity.kt

```diff
@@ -0,0 +1,129 @@
+import pm.gnosis.heimdall.utils.InfoTipDialogBuilder
+import pm.gnosis.heimdall.utils.handleQrCodeActivityResult
+import pm.gnosis.heimdall.utils.underline
+import pm.gnosis.model.Solidity
+import pm.gnosis.svalinn.common.utils.snackbar
+import pm.gnosis.svalinn.common.utils.toast
+import pm.gnosis.svalinn.common.utils.visible
+import pm.gnosis.utils.asEthereumAddress
+import pm.gnosis.utils.asEthereumAddressString
+import timber.log.Timber
+import kotlinx.android.synthetic.main.include_transfer_summary.transfer_data_fees_error as feesError
+import kotlinx.android.synthetic.main.include_transfer_summary.transfer_data_fees_info as feesInfo
+import kotlinx.android.synthetic.main.include_transfer_summary.transfer_data_fees_value as feesValue
+import kotlinx.android.synthetic.main.include_transfer_summary.transfer_data_safe_balance_after_value as balanceAfterValue
+import kotlinx.android.synthetic.main.include_transfer_summary.transfer_data_safe_balance_before_value as balanceBeforeValue
+import kotlinx.android.synthetic.main.screen_2fa_connect_start.extension_back_arrow as backArrow
+import kotlinx.android.synthetic.main.screen_2fa_connect_start.extension_coordinator as coordinator
+import kotlinx.android.synthetic.main.screen_2fa_connect_start.extension_bottom_panel as bottomPanel
+import kotlinx.android.synthetic.main.screen_2fa_connect_start.extension_progress_bar as progressBar
+
+class Connect2FaStartActivity : ViewModelActivity<PairingAuthenticatorContract>() {
+
+    override fun layout() = R.layout.screen_2fa_connect_start
+
+    override fun inject(component: ViewComponent) = viewComponent().inject(this)
+
+    override fun screenId() = ScreenId.CONNECT_2FA_START
+
+    private lateinit var safe: Solidity.Address
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        safe = intent.getStringExtra(EXTRA_SAFE_ADDRESS)?.asEthereumAddress() ?: run {
+            finish()
+            return
+        }
+
+        bottomPanel.disabled = true
+
+
+        viewModel.setup(safe)
+
+        viewModel.observableState.observe(this, Observer {
+
+            when(it) {
+
+                is PairingAuthenticatorContract.ViewUpdate.Balance -> {
+
+                    if (it.sufficient) {
+                        bottomPanel.disabled = false
+                        feesError.visibility = View.GONE
+                    } else {
+                        bottomPanel.disabled = true
+                        feesError.text = getString(R.string.insufficient_funds_please_add, it.tokenSymbol)
+                        feesError.visibility = View.VISIBLE
+                    }
+                    progressBar.visibility = View.GONE
+                    balanceBeforeValue.text = it.balanceBefore
+                    feesValue.text = it.fee.underline()
+                    balanceAfterValue.text = it.balanceAfter
+                }
+            }
+        })
+
+        lifecycleScope.launchWhenStarted {
```

Comment:
Is this scope even required here? `viewModel.estimate()` internally uses another coroutine, so this scope might not have any effect